addon/lib/Feature.jsm

```diff
@@ -1,140 +1,616 @@
+    const numMeasurements = incrementIntPref(this._studyPrefs.PerformedMeasurements.name);
+    const isFinalMeasurement = numMeasurements >= NUM_MEASUREMENTS;
+    await this._generateAndSendPing(data, isFinalMeasurement ? "final" : "progress");
+
+    // Terminate the study if we gathered enough measurements.
+    if (!isFinalMeasurement) {
+      return;
+    }
+
+    // Terminate this study!
+    this._studyUtils.endStudy({ reason: "ended-neutral" });
+  }
+
+  /**
+   * Handles errors encountered when performing the measurement.
+   */
+  async _handleError(data) {
+    this._log.error(`_handleError - reason ${data.reason}`);
+    this._isMeasuring = false;
+
+    // Send any partial data that we have.
+    const partialData = data.partial || {};
+    let payload = Cu.cloneInto(partialData, {});
+    payload.type = data.type;
+    await this._generateAndSendPing(payload, "error");
+
+    // Terminate this study!
+    this._studyUtils.endStudy({ reason: "ended-negative" });
+  }
+
+  /**
+   * Handles a message coming from the content.
+   */
+  receiveMessage(message) {
+    this._log.debug("receiveMessage - message received");
+    if (!message ||
+        !("data" in message) ||
+        !("name" in message) ||
+        typeof message.name !== "string" ||
+        typeof message.data !== "object") {
+      this._log.error("receiveMessage - received a malformed message.");
+      return;
+    }
+
+    // Handle the incoming message
+    const name = message.data.name;
+    const data = message.data.data;
+    this._log.debug(`receiveMessage - handling ${name}`);
+    switch (name) {
+      case "measurement-done":
+        this._handleMeasurementComplete(data);
+        break;
+      case "error":
+        this._handleError(data);
+        break;
+      default:
+        this._log.error(`receiveMessage - unexpected message '${name}' received`);
+    }
+  }
+
+  /**
+   * Handles the idle notification.
+   */
+  observe(subject, topic, data) {
+    this._log.trace(`observe - topic: ${topic}`);
+    if (topic !== "idle") {
+      // We're just looking for the "idle" topic here.
+      return;
+    }
+
+    this._performMeasurement();
+  }
+
+  /**
+   * Cleanup the internal frame before shutting down.
+   */
+  _cleanupFrame() {
+    // Uninstall the listener.
+    this._frame.messageManager.removeMessageListener("icqStudyMsg", this._chromeHandler);
+    this._frame = null;
+    this._chromeHandler = null;
+
+    // Dispose of the hidden browser.
+    if (this._browser !== null) {
+      this._browser.remove();
+      this._browser = null;
+    }
+
+    if (this._hiddenFrame) {
+      this._hiddenFrame.destroy();
+      this._hiddenFrame = null;
+    }
+  }
+
+  /**
+   * Called if the user disables the study or it gets uninstalled.
+   */
+  shutdown() {
+    this._log.debug("shutdown");
+
+    // Clean up the timer and the idle observer.
+    clearTimeout(this._startupTimer);
+
+    if (typeof this._idleTimeS === "number") {
```

Comment:
Good point, it was to prevent failing when shutting down if the observer wasn't already registered. There's no point to it, as I can simply check if `this._idleTimeS` contains any value.